dev-tool/src/list_net/convert.rs

```diff
@@ -51,27 +54,100 @@ impl ConvertCmd {
+                Ok(0),
+                |acc, entry_res| -> Result<usize, Error> {
+                    entry_res?;
+                    acc.map(|acc| acc + 1)
+                },
+            )
+        })?;
+
+        let storage = Arc::new(Mutex::new(InMemorySamples::with_sample_capacity(nr_users)));
+
+        let progress_bar = ProgressBar::new(nr_users.try_into().unwrap()).with_style(
+            ProgressStyle::default_bar()
+                .template("Creating Storage: [{bar:43.green}] {percent:>3}% ({pos:>5}/{len:>5})")
+                .progress_chars("=> "),
+        );
+        let bar = progress_bar.clone();
+        list_csv_files(&from_soundgarden)?
+            .par_bridge()
+            .try_for_each_with(storage.clone(), move |storage, res| {
+                res.map_err(Error::from).and_then(|path| {
+                    trace!("Processing {:?}.", path.file_name().unwrap());
+                    let history = load_history(&path)?;
+                    let new_samples =
+                        list_net_training_data_from_history(&history).with_context(|| {
+                            format!(
+                                "Invariants in soundgarden user-df broken for input file: {}",
+                                path.display()
+                            )
+                        })?;
+                    let new_samples = InMemorySamples::prepare_samples(new_samples)?;
+                    let mut storage = storage.lock().unwrap();
+                    storage.add_prepared_samples(new_samples);
+                    drop(storage);
+                    bar.inc(1);
+                    Ok(())
+                })
+            })?;
+
+        progress_bar.finish();
+
+        let storage = storage
+            .try_lock()
+            .expect("Lock was leaked by rayon or poisoned.");
+
+        progress_spin_until_done("Writing storage to disk", || {
+            storage.serialize_into_file(to_samples)
+        })?;
+        Ok(NO_ERROR)
+    }
+}
 
-            debug!("Loading history for {:?}.", path.file_name().unwrap());
-            let history = load_history(path)?;
-            debug!("Processing history");
-            for (inputs, target_prob_dist) in list_net_training_data_from_history(&history)? {
-                storage.add_sample(inputs.view(), target_prob_dist.view())?;
-            }
+fn the_first_error_has_not_been_returned_predicate<T, E>() -> impl FnMut(&Result<T, E>) -> bool {
+    let mut exit_next = false;
+    move |val| {
+        if exit_next {
+            false
+        } else {
+            exit_next = val.is_err();
+            true
         }
+    }
+}
 
-        debug!("Write output file.");
-        storage.serialize_into_file(to_samples)?;
+fn list_csv_files(
+    dir: impl AsRef<Path>,
+) -> Result<impl Iterator<Item = Result<PathBuf, Error>>, Error> {
+    let err_msg = "Reading dir with user data-frames failed.";
+    let iter = fs::read_dir(&dir)
+        .context(err_msg)?
```

Comment:
Can't we just have string here and remove `err_msg`?

Reply:
Given how the code is yes,

but the code should have (and did at some point) use the error message for
the case `entry?` or `file_type()` fail...